### DISCUSSION TITLE
unable to import and use `three.js` in `observable js` code block

### DISCUSSION THREAD

**Discussion by hopezh** [quarto-dev/quarto-cli#9392] [Q&A]
### Description

I'm unable to run [observable js](https://quarto.org/docs/interactive/ojs/) code block in a `quarto` notebook [here](https://github.com/hopezh/oatNotes/blob/main/observable_js.qmd).  

# My `{ojs}` code block:

([Reference](https://github.com/hopezh/oatNotes/blob/main/observable_js.qmd) on  why using `require()`) 

```js
<!-- import * as THREE from "npm:three"; -->
THREE = require("three");

const height = 300;
const camera = new THREE.PerspectiveCamera(70, width / height, 0.01, 10);
const scene = new THREE.Scene();
const geometry = new THREE.BoxGeometry(0.2, 0.2, 0.2);
const material = new THREE.MeshNormalMaterial();
const mesh = new THREE.Mesh(geometry, material);
const renderer = new THREE.WebGLRenderer({antialias: true});
camera.position.z = 1;
scene.add(mesh);

const animation = (time) => {
  mesh.rotation.x = time / 2000;
  mesh.rotation.y = time / 1000;
  renderer.render(scene, camera);
};

renderer.setSize(width, height);
renderer.setAnimationLoop(animation);
display(renderer.domElement);
```

# Error I got:
`OJS Syntax Error (line 76, column 1)Unexpected token
`

![2024-04-17 19 51 18](https://github.com/quarto-dev/quarto-cli/assets/23435810/6866cc90-c743-4670-a6d2-fe42055718bc)

I'm not sure why this is not working, because the same code works in the example project built with `observable framework` as discussed [here](https://talk.observablehq.com/t/unable-to-import-three-installed-via-npm-in-markdown-file-of-observable-project/9154/2). 

![0fb86757a8afd6130845c67419b92c2881c29098](https://github.com/quarto-dev/quarto-cli/assets/23435810/8cc64648-8172-4f4a-804f-7bbacc2c2693)

**Comment by mcanouil** [comment]
Prefer to share your example as plain text for future readers, unless you intend to keep the file forever <https://github.com/hopezh/oatNotes/blob/main/observable_js.qmd>.

Also please use meaningful text for your link, *e.g.*, "here" and "Reference" don't mean anything and are actually the same thing. See <https://www.w3.org/WAI/tips/writing/#make-link-text-meaningful>.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by mcanouil** [reply]
This being said, you need to simplify your code to debug it as you would do in any language.
Since I do not have NPM:

This works.

````qmd
---
title: "Penguins"
format: html
---

```{ojs}
THREE = import("https://cdnjs.cloudflare.com/ajax/libs/three.js/0.163.0/three.module.min.js");
```
````

This does not.

````qmd
---
title: "Penguins"
format: html
---

```{ojs}
const height = 300;
```
````

This does.

````qmd
---
title: "Penguins"
format: html
---

```{ojs}
{
  const height = 300;
}
```
````

I am not an expert in JavaScript/ObservableHQ.
CC @cscheid

**Reply by hopezh** [reply]
Thanks, @mcanouil. 

The following `observable js` code block in a `qmd` file works for `three.js` which is copied from the [observable notebook template for three.js](https://observablehq.com/@observablehq/three-js): 

````
---
title: "test observable.js"
format:
  html:
    toc: false
    echo: false
    keep-hidden: true
    code-tools: true
---

```{ojs}
THREE = {
  const THREE = window.THREE = await require("three@0.130.0/build/three.min.js");
  await require("three@0.130.0/examples/js/controls/OrbitControls.js").catch(() => {});
  return THREE;
}

height = 600

cube = {
  const material = new THREE.MeshNormalMaterial();
  const geometry = new THREE.BoxGeometry(1, 1, 1);
  const cube = new THREE.Mesh(geometry, material);
  return cube;
}

scene = {
  const scene = new THREE.Scene();
  scene.background = new THREE.Color(0x001b42);
  scene.add(cube);
  return scene;
}

camera = {
  const fov = 45;
  const aspect = width / height;
  const near = 1;
  const far = 1000;
  const camera = new THREE.PerspectiveCamera(fov, aspect, near, far);
  camera.position.set(2, 2, -2)
  camera.lookAt(new THREE.Vector3(0, 0, 0));
  return camera;
}

renderer = {
  const renderer = new THREE.WebGLRenderer({antialias: true});
  renderer.setSize(width, height);
  renderer.setPixelRatio(devicePixelRatio);
  const controls = new THREE.OrbitControls(camera, renderer.domElement);
  controls.addEventListener("change", () => renderer.render(scene, camera));
  invalidation.then(() => (controls.dispose(), renderer.dispose()));
  return renderer;
}

// Continuously updates
{
  while (true) {
    cube.rotation.z += 0.01;
    renderer.render(scene, camera);
    yield null;
  }
}

renderer.domElement
```
````

![2024-04-17 22 07 03](https://github.com/quarto-dev/quarto-cli/assets/23435810/1b9377f9-1305-43fe-87f8-5e3e44d660bd)

**Comment by hopezh** [comment]
The following also works, which follows the structure of the [observable notebook template for three.js](https://observablehq.com/@observablehq/three-js). 

However, **`code line numbers` is not working**...


````
---
title: "Test Observable.js"
format:
  html:
    code-block-bg: true
    code-block-border-left: true
    code-copy: true
    code-fold: false
    code-line-numbers: true
    code-tools: true
---

```{ojs}
renderer.domElement
```

```{ojs}
// Continuously updates
{
  while (true) {
    cube.rotation.z += 0.01;
    renderer.render(scene, camera);
    yield null;
  }
}
```

```{ojs}
height = 600
```

```{ojs}
cube = {
  const material = new THREE.MeshNormalMaterial();
  const geometry = new THREE.BoxGeometry(1, 1, 1);
  const cube = new THREE.Mesh(geometry, material);
  return cube;
}
```

```{ojs}
scene = {
  const scene = new THREE.Scene();
  scene.background = new THREE.Color(0x001b42);
  scene.add(cube);
  return scene;
}
```

```{ojs}
camera = {
  const fov = 45;
  const aspect = width / height;
  const near = 1;
  const far = 1000;
  const camera = new THREE.PerspectiveCamera(fov, aspect, near, far);
  camera.position.set(2, 2, -2)
  camera.lookAt(new THREE.Vector3(0, 0, 0));
  return camera;
}
```

```{ojs}
renderer = {
  const renderer = new THREE.WebGLRenderer({antialias: true});
  renderer.setSize(width, height);
  renderer.setPixelRatio(devicePixelRatio);
  const controls = new THREE.OrbitControls(camera, renderer.domElement);
  controls.addEventListener("change", () => renderer.render(scene, camera));
  invalidation.then(() => (controls.dispose(), renderer.dispose()));
  return renderer;
}
```

```{ojs}
THREE = {
  const THREE = window.THREE = await require("three@0.130.0/build/three.min.js");
  await require("three@0.130.0/examples/js/controls/OrbitControls.js").catch(() => {});
  return THREE;
}
```
````

![2024-04-17 22 24 41](https://github.com/quarto-dev/quarto-cli/assets/23435810/29d278d9-cde1-4ef6-80f8-efe9b6379c9d)

**Comment by cscheid** [comment]
````
---
title: "Penguins"
format: html
---

```{ojs}
const height = 300;
```
````

That's an invalid OJS code cell declaration:

<img width="377" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/518a2557-4aac-42b6-83d8-4eaa16fb0fc0">

OJS is not Javascript. Observable Framework is not the same as Quarto, especially when you use their pure-javascript code.
